File: klotski.py
#!/usr/local/bin/python3

""" klotski.py
Solving the Klotski block puzzle through random brute-force search.
Author: Jorge Yanar
"""

import numpy as np


def print_board(board):
  print('')
  for i in range(5):
    print('\t', end='')
    for j in range(4):
      print(board[i,j], end=' ')
    print('')
  print('')


def get_next_states(board):
  """ Returns a list of possible next states, given a board state. """
  states = []

  # Get locations of 0s in (i,j) tuples
  zeros_idx = np.where(board == 0)
  zeros_ij = list(zip(zeros_idx[0], zeros_idx[1]))

  # For each 0, check which pieces touch it
  pieces = []
  for i, j in zeros_ij:
    if i != 0 and board[i-1, j] != 0:
      pieces.append(get_board_piece_at_ij(board, (i-1, j)))
    if i != 4 and board[i+1, j] != 0:
      pieces.append(get_board_piece_at_ij(board, (i+1, j)))
    if j != 0 and board[i, j-1] != 0:
      pieces.append(get_board_piece_at_ij(board, (i, j-1)))
    if j != 3 and board[i, j+1] != 0:
      pieces.append(get_board_piece_at_ij(board, (i, j+1)))

  # Check if pieces that were touching empty space can move
  for p in pieces:
    if is_valid_move(board, p, (-1, 0)): # UP
      states.append(move_piece(board, p, (-1, 0)))
    if is_valid_move(board, p, (1, 0)):  # DOWN
      states.append(move_piece(board, p, (1, 0)))
    if is_valid_move(board, p, (0, -1)): # LEFT
      states.append(move_piece(board, p, (0, -1)))
    if is_valid_move(board, p, (0, 1)):  # RIGHT
      states.append(move_piece(board, p, (0, 1)))
  return states


def is_valid_move(board, piece, direction):
  """ Checks if a given move is valid. """
  p, i, j = piece
  deltai, deltaj = direction
  if (i == 4 and direction == (1, 0) or
      i == 0 and direction == (-1, 0) or
      j == 0 and direction == (0, -1) or
      j == 3 and direction == (0, 1)):
    return False
  if (p == 2 and i == 3 and direction == (1, 0) or
      p == 3 and j == 2 and direction == (0, 1) or
      p == 4 and i == 3 and direction == (1, 0) or
      p == 4 and j == 2 and direction == (0, 1)):
    return False
  else:
    # Single square
    if p == 1:
      return board[i+deltai, j+deltaj] == 0

    # Vertical rectangle
    elif p == 2:
      # Left/Right
      if direction == (0, -1) or direction == (0, 1):
        return (board[i,   j+deltaj] == 0 and
                board[i+1, j+deltaj] == 0)
      # Up/Down
      elif direction == (-1, 0): # UP
        return board[i+deltai, j] == 0
      elif direction == (1, 0):  # DOWN
        return board[i+1+deltai, j] == 0

    # Horizontal rectangle
    elif p == 3:
      # Up/Down
      if direction == (-1, 0) or direction == (1, 0):
        return board[i+deltai, j] == 0 and board[i+deltai, j+1] == 0
      # Left/Right
      elif direction == (0, -1):
        return board[i, j+deltaj] == 0
      elif direction == (0, 1):
        return board[i, j+1+deltaj] == 0

    # Large square block
    elif p == 4:
      # Up/Down
      if direction == (-1, 0):   # UP
        return (board[i+deltai,   j] == 0 and
                board[i+deltai, j+1] == 0)
      elif direction == (1, 0):  # DOWN
        return (board[i+1+deltai,   j] == 0 and
                board[i+1+deltai, j+1] == 0)
      # Left/Right
      elif direction == (0, -1): # LEFT
        return (board[i,   j+deltaj] == 0 and
                board[i+1, j+deltaj] == 0)
      elif direction == (0, 1):  # RIGHT
        return (board[i,   j+1+deltaj] == 0 and
                board[i+1, j+1+deltaj] == 0)


def move_piece(board, piece, direction, check_validity=False):
  """ Moves a piece on the board in the specified direction, optionally
  checking if the move is valid prior to doing so.
  """
  board = board.copy()
  p, i, j = piece
  deltai, deltaj = direction
  if check_validity and is_valid_move(board, piece, direction) == False:
    return None
  else:
    if p == 1:
      board[i+deltai, j+deltaj] = 1
      board[i, j] = 0

    elif p == 2:
      board[i,   j] = 0
      board[i+1, j] = 0
      board[i+deltai,   j+deltaj] = 2
      board[i+1+deltai, j+deltaj] = 2

    elif p == 3:
      board[i,   j] = 0
      board[i, j+1] = 0
      board[i+deltai,   j+deltaj] = 3
      board[i+deltai, j+1+deltaj] = 3

    elif p == 4:
      board[i,     j] = 0
      board[i,   j+1] = 0
      board[i+1,   j] = 0
      board[i+1, j+1] = 0
      board[i+deltai,     j+deltaj] = 4
      board[i+deltai,   j+1+deltaj] = 4
      board[i+1+deltai,   j+deltaj] = 4
      board[i+1+deltai, j+1+deltaj] = 4
  return board


def get_board_vector_repr(board, x_type_pieces=None):
  """ Returns list of 3-tuples representing pieces and their locations
  from a matrix representation of a board state. For example, a
  vertical rectangle piece (represented by 2s) at location [1, 0] would
  be represented as (2, 1, 0). The [1, 0] refers to the i,j coordinates
  of the piece's top-left corner.

  TODO: Optimize using np.where
  """
  board = board.copy()
  pieces = []
  for i in range(5):
    for j in range(4):
      if board[i,j] == 1:
        pieces.append((1, i, j))

      if board[i,j] == 2:
        pieces.append((2, i, j))
        board[i+1,j] = 0

      if board[i,j] == 3:
        pieces.append((3, i, j))
        board[i,j+1] = 0

      if board[i,j] == 4:
        pieces.append((4, i, j))
        board[i+1,j] = 0
        board[i,j+1] = 0
        board[i+1,j+1] = 0
  return pieces


def get_board_matrix_repr(pieces):
  """ Returns the 5x4 matrix representation of the board from a list of
  3-tuples representing pieces and their locations.
  """
  board = np.zeros((5, 4))
  for p in pieces:
    if p == ():
      pass
    else:
      piece = p[0]
      i, j = p[1], p[2]
      if piece == 1:
        board[i, j] = 1
      if piece == 2:
        board[i,   j] = 2
        board[i+1, j] = 2
      if piece == 3:
        board[i, j] = 3 ; board[i, j+1] = 3
      if piece == 4:
        board[i,   j] = 4 ; board[i,   j+1] = 4
        board[i+1, j] = 4 ; board[i+1, j+1] = 4
  return board


def get_board_pieces_of_type_a(board, a):
  # TODO: Faster if we do the np.where trick?
  board = board.copy()
  pieces = []
  for i in range(5):
    for j in range(4):
      if board[i,j] == a:
        pieces.append((a, i, j))
        if a == 2: board[i+1,j] = 0
        if a == 3: board[i,j+1] = 0
        if a == 4:
          board[i+1,j] = 0
          board[i,j+1] = 0
          board[i+1,j+1] = 0
  return pieces


# Possibly merge these two ^  v ?
def get_board_piece_at_ij(board, ij_tuple):
  """ Given a board and i,j, returns the vector representation
  of the piece at that position.
  """
  i, j = ij_tuple

  if board[i,j] == 0: return ()
  elif board[i,j] == 1: return (1, i, j)
  elif board[i,j] == 2:
    if i == 0: return (2, i, j)
    elif i == 4: return (2, i-1, j)
    else:
      candidate_pieces = list(
        filter(lambda p: p[2] == j,
               get_board_pieces_of_type_a(board, 2))
      )
      for p in candidate_pieces:
        if p[1] == i:
          return (2, i, j)
        elif p[1]+1 == i:
          return (2, i-1, j)
        elif p[1]-1 == i:
          return (2, i+1, j)
  elif board[i,j] == 3:
    if j == 3:
      return (3, i, j-1)
    elif j == 0:
      return (3, i, j)
    else:
      three_idx = np.where(board == 3)
      return (3, three_idx[0][0], three_idx[1][0])
  elif board[i,j] == 4:
    return get_board_pieces_of_type_a(board, 4)[0]
  return None


def remove_cycles(state_sequence):#, third=True):
  """ Removes cycles from a given sequence of states. """
  states_hashed = np.array([hash(state.tostring())
                            for state in state_sequence])
  unique_states = np.unique(states_hashed)
  # if third:
  #   iter_unique = len(unique_states)//3;
  # else:
  #   iter_unique = len(unique_states)
  for i in range(len(unique_states)):
    curr_state_idxs = np.where(states_hashed == unique_states[i])[0]
    if len(curr_state_idxs) > 1:
      bad_idx = np.arange(curr_state_idxs[0]+1, curr_state_idxs[-1]+1)
      states_hashed = np.delete(states_hashed, bad_idx)
      state_sequence = np.delete(state_sequence, bad_idx, 0)
  return state_sequence


def find_solution_path(board, solns):
  """ Runs brute-force search on a board to find a path from the given
  board state to one that satisfies the piece positions found in solns.
  """
  nvisited = 0
  states_sequence = []
  while True:
    states_sequence.append(board)
    if nvisited % 10000 == 0:
      print('States visited: {}'.format(nvisited))
      states_sequence = list(remove_cycles(states_sequence))

    nvisited += 1
    next_states = get_next_states(board)
    piece_posns = [get_board_vector_repr(n) for n in next_states]
    piece_posns = [piece for subpiece_list in piece_posns for piece in subpiece_list]
    if any([s in piece_posns for s in solns]):
      for state in next_states:
        piece_posns = get_board_vector_repr(state)
        if any([s in piece_posns for s in solns]):
          states_sequence.append(state)
          board = state ; break

      print('Found solution state!')
      states_sequence = list(remove_cycles(states_sequence))
      break
    else:
      board = next_states[np.random.randint(0, len(next_states))]
  return board, states_sequence, nvisited


###############################################################################

def main():

  # Starting state
  board = np.array([
    [2, 4, 4, 2],
    [2, 4, 4, 2],
    [2, 3, 3, 2],
    [2, 1, 1, 2],
    [1, 0, 0, 1],
  ])

  print('Starting position:')
  print_board(board)

  # Example state, right before solving
  # board2 = np.array([
  #   [2, 2, 2, 2],
  #   [2, 2, 2, 2],
  #   [1, 1, 4, 4],
  #   [1, 1, 4, 4],
  #   [3, 3, 0, 0]
  # ])

  solns = [(4, 3, 0), (4, 3, 1), (4, 3, 2)]
  solved_board, states_sequence, nvisited = find_solution_path(board, solns)

  print('Solved board:')
  print_board(solved_board)

  print('Total number of states visited: {}'.format(nvisited))
  print('Final length of state sequence, after removing cycles: {}'.format(\
                                                        len(states_sequence)))
  np.save('klotski_path.npy', states_sequence)


if __name__ == '__main__':
  main()

File: klotski.jl
#!/usr/local/bin/julia

#= klotski.jl
Solving the Klotski block puzzle through random brute-force search.
Author: Jorge Yanar
=#
using JLD

function print_board(board)
  println()
  for i = 1 : 5
    print("\t")
    for j = 1 : 4
      print(board[i,j] , " ")
    end
    println()
  end
  println()
end


function get_next_states(board)
  # Returns a list of possible next states, given board state.
  states = [];
  # Get locations of 0s in (i,j) tuples
  zeros_idx = findall(x->x==0, board);
  # For each 0, check which pieces touch it
  pieces = [];
  for idx in zeros_idx
    i = idx[1] ; j = idx[2];
    if i != 1 && board[i-1, j] != 0
      push!(pieces , get_board_piece_at_ij(board, [i-1, j]));
    end
    if i != 5 && board[i+1, j] != 0
      push!(pieces , get_board_piece_at_ij(board, [i+1, j]));
    end
    if j != 1 && board[i, j-1] != 0
      push!(pieces , get_board_piece_at_ij(board, [i, j-1]));
    end
    if j != 4 && board[i, j+1] != 0
      push!(pieces , get_board_piece_at_ij(board, [i, j+1]));
    end
  end
  pieces = unique(pieces);

  # Check if pieces that were touching empty space can move
  for p in pieces
    if is_valid_move(board, p, [-1, 0]) # UP
      push!(states, move_piece(board, p, (-1, 0)));
    end
    if is_valid_move(board, p, [1, 0])  # DOWN
      push!(states, move_piece(board, p, (1,  0)));
    end
    if is_valid_move(board, p, [0, -1]) # LEFT
      push!(states, move_piece(board, p, (0, -1)));
    end
    if is_valid_move(board, p, [0, 1])  # RIGHT
      push!(states, move_piece(board, p, (0,  1)));
    end
  end
  return states
end


function is_valid_move(board, piece, direction)
  # Checks if a given move is valid
  p, i, j = piece;
  di, dj = direction;
  # Check if movement is out of bounds
  if (i == 5 && di == 1)  ||
     (i == 1 && di == -1) ||
     (j == 1 && dj == -1) ||
     (j == 4 && dj == 1)
    return false
  end
  # Check if movement is out of bounds for specific piece type
  if (p == 2 && i == 4 && di == 1) ||
     (p == 3 && j == 3 && dj == 1) ||
     (p == 4 && i == 4 && di == 1) ||
     (p == 4 && j == 3 && dj == 1)
    return false
  else
    # Single square
    if p == 1
      return board[i+di, j+dj] == 0

    # Vertical rectangle
    elseif p == 2
      # Up/Down
      if     di == -1 return board[i+di,   j] == 0
      elseif di ==  1 return board[i+1+di, j] == 0
      # Left/Right
      elseif dj == -1 || dj == 1
        return board[i, j+dj] == 0 && board[i+1, j+dj] == 0
      end

    # Horizontal rectangle
    elseif p == 3
      # Up/Down
      if di != 0 return board[i+di, j] == 0 && board[i+di, j+1] == 0 end
      # Left/Right
      if     dj == -1 return board[i, j+dj]   == 0
      elseif dj ==  1 return board[i, j+1+dj] == 0
      end

    # Large square block
    elseif p == 4
      # Up/Down
      if     di == -1 return board[i+di,   j] == 0 && board[i+di,   j+1] == 0
      elseif di ==  1 return board[i+1+di, j] == 0 && board[i+1+di, j+1] == 0
      # Left/Right
      elseif dj == -1 return board[i,   j+dj] == 0 && board[i+1,   j+dj] == 0
      elseif dj ==  1 return board[i, j+1+dj] == 0 && board[i+1, j+1+dj] == 0
      end
    end
  end
end


function move_piece(board, piece, direction, check_validity=false)
  # Moves a piece on the board in the specified direction, optionally
  # checking if the move is valid prior to doing so.
  board = copy(board);
  p, i, j = piece;
  di, dj = direction;
  if check_validity && is_valid_move(board, piece, direction) == false
    return nothing
  else
    if p == 1
      board[i+di, j+dj] = 1;
      board[i, j] = 0;

    elseif p == 2
      board[i,   j] = 0;
      board[i+1, j] = 0;
      board[i+di,   j+dj] = 2;
      board[i+1+di, j+dj] = 2;

    elseif p == 3
      board[i,   j] = 0;
      board[i, j+1] = 0;
      board[i+di,   j+dj] = 3;
      board[i+di, j+1+dj] = 3;

    elseif p == 4
      board[i,     j] = 0;
      board[i,   j+1] = 0;
      board[i+1,   j] = 0;
      board[i+1, j+1] = 0;
      board[i+di,     j+dj] = 4;
      board[i+di,   j+1+dj] = 4;
      board[i+1+di,   j+dj] = 4;
      board[i+1+di, j+1+dj] = 4;
    end
  end
  return board
end


function get_board_vector_repr(board)
  board = copy(board);
  pieces = [];
  for i = 1 : 5
    for j = 1 : 4
      if board[i,j] == 1
        push!(pieces, (1, i, j));
      end
      if board[i,j] == 2
        push!(pieces, (2, i, j));
        board[i+1,j] = 0;
      end
      if board[i,j] == 3
        push!(pieces, (3, i, j));
        board[i,j+1] = 0;
      end
      if board[i,j] == 4
        push!(pieces, (4, i, j));
        board[i+1,  j] = 0;
        board[i,  j+1] = 0;
        board[i+1,j+1] = 0;
      end
    end
  end
  return pieces
end


function get_board_matrix_repr(pieces)
  board = zeros((5, 4));
  for p in pieces
    piece = p[1];
    i, j = p[2], p[3];
    if piece == 1
      board[i, j] = 1;
    elseif piece == 2
      board[i,   j] = 2;
      board[i+1, j] = 2;
    elseif piece == 3
      board[i,   j] = 3;
      board[i, j+1] = 3;
    elseif piece == 4
      board[i,   j] = 4; board[i,   j+1] = 4;
      board[i+1, j] = 4; board[i+1, j+1] = 4;
    end
  end
  return board
end


function get_board_pieces_of_type_a(board, a)
  board = copy(board);
  pieces = [];
  for i = 1 : 5
    for j = 1 : 4
      if board[i,j] == a
        push!(pieces, (a, i, j));
        if a == 2 board[i+1,j] = 0 end
        if a == 3 board[i,j+1] = 0 end
        if a == 4
          board[i+1,   j] = 0;
          board[i,   j+1] = 0;
          board[i+1, j+1] = 0;
        end
      end
    end
  end
  return pieces
end


function get_board_piece_at_ij(board, ij_tuple)
  i, j = ij_tuple;
  if     board[i,j] == 0 return ()
  elseif board[i,j] == 1 return (1, i, j)
  elseif board[i,j] == 2
    if     i == 1 return (2, i, j)
    elseif i == 5 return (2, i-1, j)
    else
      # Need to figure out which 2-piece this belongs to
      candidates_2s = filter(p->p[3] == j, get_board_pieces_of_type_a(board, 2));
      for p in candidates_2s
        if     p[2]   == i return (2, i, j)
        elseif p[2]+1 == i return (2, i-1, j)
        elseif p[2]-1 == i return (2, i+1, j)
        end
      end
    end
  elseif board[i,j] == 3
    if     j == 4 return (3, i, j-1)
    elseif j == 1 return (3, i, j)
    else
      three_idx = findall(x->x==3, board);
      return (3, three_idx[1][1], three_idx[1][2])
    end
  elseif board[i,j] == 4
    return get_board_pieces_of_type_a(board, 4)[1]
  end
  return nothing
end


function remove_cycles(states_sequence)
  # """ Removes cycles from a given sequence of states. """
  states_hashed = map(v -> hash(v), states_sequence);
  unique_states = unique(states_hashed);
  for i = 1 : length(unique_states)
    curr_state_idxs = findall(hv->hv==unique_states[i], states_hashed);
    if length(curr_state_idxs) > 1
      # Get rid of all repeats
      bad_idx = curr_state_idxs[1]+1 : curr_state_idxs[end];
      deleteat!(states_hashed, bad_idx);
      deleteat!(states_sequence, bad_idx);
    end
  end
  return states_sequence
end


function find_solution_path(board, solns)
  nvisited = 1;
  states_seq = [board];
  while true
    # Generate list of possible next states
    next_states = get_next_states(board);

    # For each state, check whether it contains a piece placed
    # in a winning configuration
    for state in next_states
      pieces = get_board_vector_repr(state);
      if any(s -> s in solns, pieces)
        println("Found solution state!");
        push!(states_seq, state);
        nvisited += 1;
        states_seq = remove_cycles(states_seq);
        return state, states_seq, nvisited
      end
    end

    # None found, randomly select one of the next_states
    board = rand(next_states);
    push!(states_seq, board);
    nvisited += 1;

    # Trim cycles every 10,000 steps to reduce computational load
    if nvisited % 10000 == 0
      print("States visited: ", nvisited);
      states_seq = remove_cycles(states_seq);
      println(" current pathlength: ", length(states_seq));
    end
  end
end

###############################################################################

function main()
  
  # Starting state
  board = [
    2 4 4 2
    2 4 4 2
    2 3 3 2
    2 1 1 2
    1 0 0 1
  ];

  println("Starting position:");
  print_board(board)

  solns = [(4, 4, 2)];
  solved_board, states_sequence, nvisited = find_solution_path(board, solns);

  println("Solved board:")
  print_board(solved_board)

  println("Total number of states visited: ", nvisited);
  println("Final length of state sequence, after removing cycles: ", length(states_sequence));

  save("klotski_path.jld", "states_sequence", states_sequence,
                              "solved_board", solved_board,
                                  "nvisited", nvisited);
end

main()
